File: src/server/src/velvet_dawn/mechanics/selectors/selector.py
from abc import ABC
from typing import List, Union, Optional

from velvet_dawn.dao.models import UnitInstance, TileInstance

from velvet_dawn.dao.models.world_instance import WorldInstance
from velvet_dawn.mechanics.selectors.filters import Filters


class Selector(ABC):
    def __init__(
            self,
            selector_name: str,
    ):
        self.full_selector = None
        self.attribute = None
        self.selector_name = selector_name
        self.filters: Filters = Filters()

        self.chained_selector: Optional[Selector] = None

    def new(self) -> 'Selector':
        raise NotImplementedError()

    def instantiate(self, full_selector, filters: List[str], attribute: str) -> 'Selector':
        new_copy = self.new()
        new_copy.full_selector = full_selector
        new_copy.attribute = attribute

        if filters:
            for filter in filters:
                if "=" in filter:
                    key, value = filter.split("=")
                    new_copy.filters.add_filter(key, value)
                else:
                    new_copy.filters.add_filter(filter)

        return new_copy

    def get_selection(
            self,
            instance: Union[TileInstance, UnitInstance, WorldInstance]
    ) -> List[Union[UnitInstance, TileInstance, WorldInstance]]:
        raise NotImplementedError()

    def get_chained_selection(self, instance: Union[TileInstance, UnitInstance, WorldInstance]) -> List[Union[UnitInstance, TileInstance, WorldInstance]]:
        direct_selection = self.get_selection(instance)

        if not self.chained_selector:
            return direct_selection

        chained_selection = set()
        for item in direct_selection:
            chained_selection.update(self.chained_selector.get_chained_selection(item))

        return list(chained_selection)

    def function_set(self, instance: Union[TileInstance, UnitInstance], value: Union[str, int, float]):
        for item in self.get_chained_selection(instance):
            item.set_attribute(self.attribute, value)

    def function_add(self, instance: Union[TileInstance, UnitInstance], value: Union[str, int, float]):
        for item in self.get_chained_selection(instance):
            print(item.get_attribute(self.attribute, default=0))
            print(value)
            item.set_attribute(self.attribute, item.get_attribute(self.attribute, default=0) + value)

    def function_subtract(self, instance: Union[TileInstance, UnitInstance], value: Union[str, int, float]):
        for item in self.get_chained_selection(instance):
            item.set_attribute(self.attribute, item.get_attribute(self.attribute, default=0) - value)

    def function_reset(self, instance: Union[TileInstance, UnitInstance], value: Union[str, int, float]):
        for item in self.get_chained_selection(instance):
            item.reset_attribute(self.attribute, value)

    def function_multiply(self, instance: Union[TileInstance, UnitInstance], value: Union[str, int, float]):
        for item in self.get_chained_selection(instance):
            item.set_attribute(self.attribute, item.get_attribute(self.attribute, default=0) * value)

    def function_add_tag(self, instance: Union[TileInstance, UnitInstance], value: Union[str, int, float]):
        for item in self.get_chained_selection(instance):
            item.add_tag(value)

    def function_remove_tag(self, instance: Union[TileInstance, UnitInstance], value: Union[str, int, float]):
        for item in self.get_chained_selection(instance):
            item.remove_tag(value)

    def function_equals(self, instance: Union[TileInstance, UnitInstance], value):
        """ Compare if the given value is equal the selectors' id / attribute value """
        instances = self.get_chained_selection(instance)
        if not instances:
            return False

        for item in instances:
            if self.attribute: 
                if item.get_attribute(self.attribute, default=None) != value:
                    return False
            elif item.entity_id != value:
                return False

        return True

    def function_less_than(self, instance: Union[TileInstance, UnitInstance], value):
        """ Compare if the entities attribute is less than a given value """
        instances = self.get_chained_selection(instance)
        if not instances or not self.attribute:
            return False

        for item in instances:
            if item.get_attribute(self.attribute, default=None) >= int(value):
                return False

        return True

    def function_less_than_equals(self, instance: Union[TileInstance, UnitInstance], value):
        """ Compare if the entities attribute is less/equal than a given value """
        instances = self.get_chained_selection(instance)
        if not instances or not self.attribute:
            return False

        for item in instances:
            if item.get_attribute(self.attribute, default=0) > int(value):
                return False

        return True

    def function_get_value(self, instance: Union[TileInstance, UnitInstance]):
        """ This function effectively works to get the mean value of the data """
        instances = self.get_chained_selection(instance)
        n_instances = len(instances)
        if n_instances == 0:
            return 0

        all_string, string_value = True, None
        all_number, number_values = True, []

        for instance in instances:
            value = instance.get_attribute(self.attribute) if self.attribute else instance.id
            if value is not None:
                if isinstance(value, str):
                    all_number = False

                    if string_value is None:
                        string_value = value
                    elif string_value != value:
                        return None

                elif isinstance(value, int) or isinstance(value, float) or isinstance(value, bool):
                    all_string = False
                    number_values.append(float(value))
                else:
                    # Invalid type
                    return None

        if all_string and all_number:
            # Must have no valid values, e.g. all [None, None]
            return None

        if all_string:
            return string_value

        if all_number:
            if number_values:
                return sum(number_values) / len(number_values)
        return None

    def function_has_tag(self, instance: Union[TileInstance, UnitInstance], value: str):
        """ Check if the selection has a given tag """
        instances = self.get_chained_selection(instance)
        if not instances:
            return False

        for item in instances:
            if not item.has_tag(value):
                return False

        return True

File: src/server/src/velvetdawn/core/mechanics/selectors/Selector.java
package velvetdawn.core.mechanics.selectors;

import velvetdawn.core.VelvetDawn;
import velvetdawn.core.models.anytype.Any;
import velvetdawn.core.models.anytype.AnyFloat;
import velvetdawn.core.models.anytype.AnyNull;
import velvetdawn.core.models.anytype.AnyString;
import velvetdawn.core.models.instances.Instance;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;

public abstract class Selector {

    protected final VelvetDawn velvetDawn;

    public Filters filters;
    public Selector chainedSelector = null;
    public String attribute = null;

    public Selector(VelvetDawn velvetDawnCore) {
        this.velvetDawn = velvetDawnCore;
        this.filters = new Filters(velvetDawn);
    }

    protected abstract Collection<Instance> getSelection(Instance instance);

    public Collection<Instance> getChainedSelection(Instance instance) {
        var directSelection = this.getSelection(instance);
        if (this.chainedSelector == null)
            return directSelection;

        var chainedSelection = new HashSet<Instance>();
        directSelection.forEach(item -> {
            chainedSelection.addAll(this.chainedSelector.getChainedSelection(item));
        });

        return chainedSelection;
    }

    public void funcSet(Instance instance, Any value) {
        this.getChainedSelection(instance).forEach(item -> {
            item.attributes.set(this.attribute, value);
        });
    }

    public void funcAdd(Instance instance, Any value) {
        this.getChainedSelection(instance).forEach(item ->
                item.attributes.set(
                        this.attribute,
                        item.attributes.get(this.attribute).add(value)));
    }

    public void funcSub(Instance instance, Any value) {
        this.getChainedSelection(instance).forEach(item ->
                item.attributes.set(
                        this.attribute,
                        item.attributes.get(this.attribute).sub(value)));
    }

    public void funcMul(Instance instance, Any value) {
        this.getChainedSelection(instance).forEach(item ->
                item.attributes.set(
                        this.attribute,
                        item.attributes.get(this.attribute).mul(value)));
    }

    public void funcReset(Instance instance, Any value) {
        this.getChainedSelection(instance).forEach(item ->
                item.attributes.reset(this.attribute, value));
    }

    public void funcAddTag(Instance instance, Any tag) {
        this.getChainedSelection(instance).forEach(item -> item.tags.add(tag.toString()));
    }

    public void funcRemoveTag(Instance instance, Any tag) {
        this.getChainedSelection(instance).forEach(item -> item.tags.remove(tag.toString()));
    }

    /** Compare if the given value is equal the selectors' id / attribute value */
    public boolean funcEquals(Instance instance, Any value) {
        var instances = this.getChainedSelection(instance);
        if (instances.isEmpty())
            return false;

        for (Instance item: instances) {
            if (this.attribute != null) {
                if (!item.attributes.get(this.attribute).equals(value))
                    return false;
            } else if (!Objects.equals(item.datapackId, value.toString()))
                return false;
        }

        return true;
    }

    /** Compare if the entities attribute is less than a given value */
    public boolean funcLessThan(Instance instance, Any value) {
        var instances = this.getChainedSelection(instance);
        if (instances.isEmpty() || this.attribute == null)
            return false;

        for (Instance item: instances) {
            if (item.attributes.get(this.attribute).gte(value))
                return false;
        }

        return true;
    }

    /** Compare if the entities attribute is less/equal than a given value */
    public boolean funcLessThanEqual(Instance instance, Any value) {
        var instances = this.getChainedSelection(instance);
        if (instances.isEmpty() || this.attribute == null)
            return false;

        for (Instance item: instances) {
            if (item.attributes.get(this.attribute).gt(value))
                return false;
        }

        return true;
    }

    /** This function effectively works to get the mean value of the data */
    public Any funcGetValue(Instance instance) {
        var instances = this.getChainedSelection(instance);
        if (instances.isEmpty())
            return Any.Null();

        boolean allString = true;
        boolean allNumber = true;
        boolean allNull = true;
        String stringValue = null;
        List<Float> numberValues = new ArrayList<>();

        for (Instance item: instances) {
            Any value = this.attribute == null
                    ? new AnyString(item.datapackId)
                    : item.attributes.get(this.attribute);

            if (value instanceof AnyNull)
                continue;

            if (value instanceof AnyString) {
                allNumber = false;
                allNull = false;

                if (stringValue == null)
                    stringValue = value.toString();
                else if (!Objects.equals(stringValue, value.toString()))
                    return Any.Null();

            }
            else if (value instanceof AnyFloat) {
                allString = false;
                allNull = false;
                numberValues.add(value.toNumber());
            } else {
                // Invalid type
                return Any.Null();
            }
        }

        if (allNull)
            return Any.Null();

        if (allString)
            return new AnyString(stringValue);

        if (allNumber && !numberValues.isEmpty()) {
            float sum = 0;
            for (float v: numberValues)
                sum += v;
            return new AnyFloat(sum / ((float) numberValues.size()));
        }

        return Any.Null();
    }

    /** Check if the selection has a given tag */
    public boolean funcHasTag(Instance instance, Any tag) {
        var instances = this.getChainedSelection(instance);
        if (instances.isEmpty())
            return false;

        for (Instance item: instances) {
            if (!item.tags.contains(tag.toString()))
                return false;
        }

        return true;
    }
}
